### PR TITLE
Allow environment variable and tag specification for models deployed via MLflow SageMaker integration

### PR DESCRIPTION
**Pull request by jhallard**
## Related Issues/PRs

N/A 

 #xxx

## What changes are proposed in this pull request?

This change serves to allow users of the MLflow SageMaker deployment a way of setting environment variables and tags for the models they deploy. 

This is accomplished by adding two new options to the deployment config dictionary, available to set during deployment creation or update

1. `env`: dictionary of key/value pairs representing environment variables to set for the created model
2. `tags`: dictionary of key/value pairs representing tags to set for the created model

For `mlflow deployments {create|update}` CLI, these are set via the `-C {<json>}` interface, similar to `vpc_config`. For the Python SDK interface, these are passed in as simple dictionaries.

The primary motivation here was that one of our models was configured via envvars and we had no simple way of setting these when deploying via the `mlflow deployments create -t sagemaker` command without going into the AWS console and performing a manual override there. We also keep track of resources via tags, and there was no simple way of setting those either.

## How is this patch tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

Updated existing unit tests for both the SDK and CLI interfaces to ensure that environment variables are threaded through correctly. I also updated the `moto` SageMaker backend to have a `list_tags` method to help test the tag-setting capabilities.

## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Users of the SageMaker deployment functionality can now configure environment variables and tags for the deployed models from both the CLI and Python SDK.

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [x] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Comment by dbczumar** [comment]
@harupy Can you take a look?

File: mlflow/sagemaker/__init__.py [hunk 1]
```diff
@@ -310,6 +312,8 @@ def _deploy(
                                     mfs.deploy(..., data_capture_config=data_capture_config)
 
     :param variant_name: The name to assign to the new production variant.
+    :param env: A optional dictionary of environment variables to set for the model.
+    :param tags: A optional dictionary of tags to apply to the endpoint.
```
Comment (harupy) [inline mlflow/sagemaker/__init__.py:316]:
```suggestion
    :param env: An optional dictionary of environment variables to set for the model.
    :param tags: An optional dictionary of tags to apply to the endpoint.
```

File: mlflow/sagemaker/__init__.py [hunk 2]
```diff
@@ -841,8 +849,8 @@ def status_check_fn():
 
         if transform_job_info["TransformJobStatus"] == "Stopping":
             return _SageMakerOperationStatus.in_progress(
-                "Termination is still in progress. Current batch transform job status:\
-                    {transform_job_status}".format(
+                "Termination is still in progress. Current batch transform job status: "
+                "       {transform_job_status}".format(
```
Comment (harupy) [inline mlflow/sagemaker/__init__.py:853]:
```suggestion
                "{transform_job_status}".format(
```

**Review by harupy — APPROVED** [review]
Left a few minor comments, otherwise looks good!

File: mlflow/sagemaker/__init__.py [hunk 3]
```diff
@@ -2366,23 +2421,27 @@ def update_deployment(
...
+                message="A model_uri must be provided when updating a SageMaker deployment",
                 error_code=INVALID_PARAMETER_VALUE,
             )
 
         if final_config["mode"] not in [DEPLOYMENT_MODE_ADD, DEPLOYMENT_MODE_REPLACE]:
             raise MlflowException(
-                message=f"Invalid mode `{final_config['mode']}` for deployment \
-                        to a pre-existing application",
+                message=(
+                    f"Invalid mode `{final_config['mode']}` for deployment                        "
```
Comment (harupy) [inline mlflow/sagemaker/__init__.py:2442]:
```suggestion
                    f"Invalid mode `{final_config['mode']}` for deployment"
```

File: tests/sagemaker/mock/__init__.py [hunk 4]
```diff
@@ -497,6 +511,14 @@ def list_models(self):
             summaries.append(summary)
         return summaries
 
+    def list_tags(self, resource_arn, region_name):
+        """
+        Modifies backend state during calls to the SageMaker "ListTags" API
+        https://docs.aws.amazon.com/sagemaker/latest/APIReference/API_ListTags.html
+        """
+        model = [model for model in self.models.values() if model.arn == resource_arn][0]
```
Comment (harupy) [inline tests/sagemaker/mock/__init__.py:519]:
```suggestion
        model = next(model for model in self.models.values() if model.arn == resource_arn)
```

**Review by harupy — APPROVED** [review]
LGTM!

**Pull request merged**